refactor(backend): ノートのエクスポート処理でStreams APIを使うように

okayurisotto opened the pull request:
## What

これまで、ノートのエクスポート処理では、`fs.createWriteStream()`で開いたファイルに対し、独自に定義した`write()`関数の呼び出しで書き込んでいました。しかしこの部分のコードは読みにくく、再利用性も低いものでした。実際、アンテナなどのノート以外のもののエクスポートでも同様の処理が行われており、この部分のコードは重複した状態にありました。

そこでこのPRでは、Web標準のStreams APIを使用して、

1. ノートの取得
2. JSON配列化
3. 文字列のエンコード
4. ファイルへの書き込み

を繋げて行うようにしました。

このうち、2と4の処理では再利用性の高いモジュールを新たに作り、3の処理にはNode.jsの標準機能（`TextEncoderStream`）を使用しました。

## Why

メンテナンス性が低かったため。

また、他の`Export*ProcessorSerivce`と共通した処理をまとめるには、Streams APIを活用するのがいいように思えたため。

## Additional info (optional)

このPRで使っているStreams APIは[Node.jsのもの](https://nodejs.org/api/stream.html)**ではなく**、[Web標準のもの](https://developer.mozilla.org/ja/docs/Web/API/Streams_API)になります。Node.jsのAPIを使いたかったのですが、とても使いづらく、TypeScriptとの相性も悪いようだったため、Web標準のものを採用することにしました。このPRで追加した`FileWriterStream`は、Node.jsのAPIである`fs.createWriteStream()`に相当する、Web標準のAPIを使用する実装です。

このPRがマージされた際には、他の`Export*ProcessorSerivce`についても同様のリファクタリングをしていく予定です。とりあえずこのPRで、リファクタリングの方向性が悪くないかレビューしていただきたいです。

元コードにあったN+1の解消はできていません。このPRでその問題まで扱う予定はありません。

`for...of`ループ内で`await`している部分があり、`Promise.all()`に書き換えることで処理速度の向上が見込めますが、サーバーへの負荷を考えて見送りました。`promise-limit`のようなパッケージを使って`Promise`の同時実行数を制限すれば負荷を抑えつつ処理を速められるかもしれませんが、未検証です。

## Checklist

- [x] Read the [contribution guide](https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md)
- [ ] Test working in a local environment
- [ ] (If needed) Update CHANGELOG.md
- [ ] (If possible) Add tests

okayurisotto committed:
refactor(backend): ノートのエクスポート処理でStreams APIを使うように

syuilo committed:
Merge branch 'develop' into pr/13465

syuilo commented:
ローカルで試したところ
```
Error: Missing key for job 1. updateProgress
    at Scripts.finishedErrors (C:\Users\syuilo\git\misskey\node_modules\.pnpm\bullmq@5.1.9\node_modules\bullmq\dist\cjs\classes\scripts.js:272:24)
    at Scripts.updateProgress (C:\Users\syuilo\git\misskey\node_modules\.pnpm\bullmq@5.1.9\node_modules\bullmq\dist\cjs\classes\scripts.js:209:24)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Job.updateProgress (C:\Users\syuilo\git\misskey\node_modules\.pnpm\bullmq@5.1.9\node_modules\bullmq\dist\cjs\classes\job.js:298:9)
```
が出て正しくエクスポートされないですね

okayurisotto committed:
fixup! refactor(backend): ノートのエクスポート処理でStreams APIを使うように

`await`忘れにより、ジョブがすぐに完了したことになり削除されてしまっていた。
それによって、`NoteStream`内での`updateProgress`メソッドの呼び出しで、`Missing key for job`のエラーが発生することがあった。

okayurisotto commented:
私の環境でもエラーを確認しました。`await`を忘れていたことが原因だと思われます。修正しましたのでお手数ですが再度確認をお願いします。

syuilo commented:
👍🏻 👍🏻
